woocommerce/woocommerce-gateway-stripe
# Update the billing address on saved card before creating the payment intent/processing payment

Fixes #1682 

## Changes proposed in this Pull Request:







When a customer saves a payment method to their account, the billing address provided on the checkout is saved to that payment method. Say this customer moves/changes addresses and proceeds through checkout using their saved payment method Stripe could block the payment due to zip code verification erros.

In Stripe, we currently update the billing address on the saved method asynchronously on an action scheduler hook, but this doesn't update the payment method address quick enough and it needs to happen before the intent is created.

This PR fixes this issue by:
1. Removing code which updates the saved payment methods billing address on an action scheduler hook
2. Updating the saved payment method before creating the payment/setup intent
3. Move the existing `update_saved_payment_method()` to the abstract Stripe gateway base class so it can be called when UPE is enabled and disabled.

## Testing instructions





I wasn't able to replicate zip code verification errors so to test these changes I just confirmed that the payment method billing address wasn't being updated at the time of processing the payment and didn't match the shipping address.

The two addresses I used for testing were:

Address A:
```
Australia
123 Brisbane Street
Brisbane City, Queensland, 4000
```

Address B:
```
Australia
123 Sydney Street
Sydney, New South Wales, 2000
```


1. Make sure UPE is enabled and also the "Enable payments via saved cards" setting is selected
2. Add a product to your cart and while on the checkout page, enter Address A
3. Enter a new card and make sure to save it to your account
4. Open Stripe and confirm the payment method has the "Brisbane Street" address:
![image](https://github.com/woocommerce/woocommerce-gateway-stripe/assets/2275145/9b8ec612-ff96-482f-a18d-7ee1bfa3322b)
5. While on `develop`, add another product the cart, visit the checkout.
6. Enter in Address B into the billing details and use the same saved card that was saved above.
7. Submit the checkout and view the new payment in the stripe dashboard.
8. Notice that the payment method still has the old customer address, but the payment itself has the new address: ![image](https://github.com/woocommerce/woocommerce-gateway-stripe/assets/2275145/3062d93e-f14b-44bd-9852-f8694de812ac)
9. Check out this branch and run through the checkout again with a different Address.
10. Confirm that the payment method billing address has been updated and matches the payment method address.

Here's a full list of tests I ran:
- [x] UPE enabled
    - [x] Existing saved card
    - [x] New saved card - shouldn't update payment method as it will be set when creating
    - [x] No saved payment method
- [x] UPE disabled
    - [x] Existing saved card
    - [x] New saved card
    - [x] No saved payment method
- [x] SEPA saved payment method



---


**Post merge**
